flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderElectionService.java

```diff
@@ -0,0 +1,219 @@
+			kubernetesWatch.close();
+		}
+	}
+
+	@Override
+	protected void writeLeaderInformation() {
+		try {
+			kubeClient.checkAndUpdateConfigMap(
+				configMapName,
+				configMap -> {
+					if (leaderElector.hasLeadership(configMap)) {
+						// Get the updated ConfigMap with new leader information
+						if (confirmedLeaderAddress != null && confirmedLeaderSessionID != null) {
+							configMap.getData().put(LEADER_ADDRESS_KEY, confirmedLeaderAddress);
+							configMap.getData().put(LEADER_SESSION_ID_KEY, confirmedLeaderSessionID.toString());
+						}
+						configMap.getLabels().putAll(configMapLabels);
+						return Optional.of(configMap);
+					}
+					return Optional.empty();
+				}).get();
+		} catch (Exception e) {
+			leaderContender.handleError(new Exception("Could not update ConfigMap " + configMapName, e));
+		}
+	}
+
+	@Override
+	protected boolean hasLeadership() {
+		return kubeClient.getConfigMap(configMapName)
+			.map(leaderElector::hasLeadership)
+			.orElse(false);
+	}
+
+	@Override
+	public String toString() {
+		return "KubernetesLeaderElectionService{configMapName='" + configMapName + "'}";
+	}
+
+	private class LeaderCallbackHandlerImpl extends KubernetesLeaderElector.LeaderCallbackHandler {
+
+		@Override
+		public void isLeader() {
+			onGrantLeadership();
+		}
+
+		@Override
+		public void notLeader() {
+			// Clear the leader information in ConfigMap
+			try {
+				kubeClient.checkAndUpdateConfigMap(
+					configMapName,
+					configMap -> {
+						// Do not need to check the leader here
+						configMap.getData().remove(LEADER_ADDRESS_KEY);
+						configMap.getData().remove(LEADER_SESSION_ID_KEY);
+						return Optional.of(configMap);
+					}
+				).get();
+			} catch (Exception e) {
+				leaderContender.handleError(
+					new Exception("Could not remove leader information from ConfigMap " + configMapName, e));
+			}
+			onRevokeLeadership();
+			// Continue to contend the leader
+			CompletableFuture.runAsync(leaderElector::run, executor);
+		}
+	}
+
+	private class ConfigMapCallbackHandlerImpl implements FlinkKubeClient.WatchCallbackHandler<KubernetesConfigMap> {
+
+		@Override
+		public void onAdded(List<KubernetesConfigMap> configMaps) {
+			// noop
+		}
+
+		@Override
+		public void onModified(List<KubernetesConfigMap> configMaps) {
+			if (hasLeadership()) {
+				configMaps.forEach(configMap -> {
+					if (isLeaderUpdatedExternally(configMap)) {
+						if (configMap.getName().equals(configMapName)) {
+							// the data field does not correspond to the expected leader information
+							if (logger.isDebugEnabled()) {
+								logger.debug("Correcting leader information in {} by {}.",
+									configMapName, leaderContender.getDescription());
+							}
+							writeLeaderInformation();
+						} else {
+							logger.warn("Ignoring the modified event since it does not belong to {}.", this);
+						}
+					}
+				});
+			}
+		}
+
+		@Override
+		public void onDeleted(List<KubernetesConfigMap> configMaps) {
```

Comment:
Should we also add an assertion on the confip map name for the delete events?

Reply:
Yes. Same as above.

Reply:
How can it happen that ConfigMaps are being deleted?

Reply:
Since the leader ConfigMap may be updated or deleted externally by the user, I introduce the `Watcher` in `KubernetesLeaderElectionService` for this situation. When it is deleted or wrongly updated, we will create a new one or correct the content.

I think you mean that we do not need to handle this situation and just throw a fatal exception and failed the Flink cluster. Right?

Reply:
I see. Hmm, thinking about it again I came to the conclusion that such a safety mechanism against accidental deletions might harden operations. I wouldn't have said that we need it for the first version but since you have it already implemented, let's keep it.

Reply:
Isn't the `KubernetesLeaderElector` responsible for creating the initial `configMap`? If this is the case, then I would assume that this component should also be responsible for ensuring that this map gets recreated in case of a deletion.

Reply:
Yes. `KubernetesLeaderElector ` is responsible for creating the leader ConfigMap. And it could recreate it again when trying to renew the lock. However, the data of new ConfigMap is empty(just the annotation is kept). Then our running jobs, job graphs will be lost. That's why I try to create a same ConfigMap in `onDeleted` callback.

Reply:
Wouldn't this be covered by the `onModified` callback which the leader will receive? The leader should then try to re-create the lost information.

This actually brings me to another question: What happens if the configMap is being deleted by the user and at the same time the current leader dies for some reason? Who would then re-create the configMap? And even if we are able to recreate the configMap, how would we be able to recover the old state (e.g. all the submitted jobs)?